### PR TITLE
Move new Delaunay code from milestone/pointcloud

### PR DESCRIPTION
**Pull request by jpolchlo**
The Delaunay triangulation code has been under development in the `milestone/pointcloud` branch, but that branch is slated to be broken off into a new project. In preparation, it's time to move the new triangulator into `master`.

Signed-off-by: jpolchlo <jpolchlopek@azavea.com>

### CODE REVIEW

File: raster/src/main/scala/geotrellis/raster/triangulation/DelaunayRasterizer.scala [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+import geotrellis.vector._
+import geotrellis.vector.triangulation._
+
+object DelaunayRasterizer {
+
+  /**
+   * Produces a Tile with shape given by a RasterExtent where pixels are linearly interpolated
+   * values of the z-coordinates of each triangle vertex.
+   */
+  def rasterize(
```
Comment (jpolchlo) [inline raster/src/main/scala/geotrellis/raster/triangulation/DelaunayRasterizer.scala]: Probably worth noting that this rasterizer is pretty specialized.  A more general interface would be desirable.

File: vector/src/main/scala/geotrellis/vector/RobustPredicates.scala [hunk 2]
```diff
@@ -0,0 +1,81 @@
+package geotrellis.vector
+
+import com.vividsolutions.jts.geom.Coordinate
+import org.apache.commons.math3.linear._
+
+import geotrellis.util.Constants.{DOUBLE_EPSILON => EPSILON}
+
+
+object RobustPredicates {
```
Comment (echeipesh) [inline vector/src/main/scala/geotrellis/vector/RobustPredicates.scala]: Needs a brief scaladoc for these functions.

File: vector/src/main/scala/geotrellis/vector/triangulation/HalfEdgeTable.scala [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+  private final val MAXSIZE: Int = Int.MaxValue / 3
+
+  // Once our buckets get 90% full, we need to resize
+  private final val FACTOR: Double = 0.9
+
+  private var limit = (size * FACTOR).toInt
+
+  private val freed = collection.mutable.ListBuffer.empty[Int]
+
+  def join(e: Int, opp: Int): Unit = {
```
Comment (echeipesh) [inline vector/src/main/scala/geotrellis/vector/triangulation/HalfEdgeTable.scala]: doc needs list of `opp`

File: vector/src/main/java/geotrellis/vector/ShewchuksDeterminant.java [hunk 4]
```diff
@@ -0,0 +1,1863 @@
...
+ *
+ * [...]
+ *  Placed in the public domain by
+ *  Jonathan Richard Shewchuk
+ *  School of Computer Science
+ *  Carnegie Mellon University
+ *  5000 Forbes Avenue
+ *  Pittsburgh, Pennsylvania  15213-3891
+ *  jrs@cs.cmu.edu
+ *  [...]
```
Comment (echeipesh) [inline vector/src/main/java/geotrellis/vector/ShewchuksDeterminant.java]: Need to create a CQ for this code.

File: vector/src/main/scala/geotrellis/vector/triangulation/DelaunayPointSet.scala [hunk 5]
```diff
@@ -0,0 +1,49 @@
+package geotrellis.vector.triangulation
+
+import com.vividsolutions.jts.geom.Coordinate
+
+import geotrellis.util.Constants.{DOUBLE_EPSILON => EPSILON}
+import geotrellis.vector.Point
+
+trait DelaunayPointSet {
```
Comment (echeipesh) [inline vector/src/main/scala/geotrellis/vector/triangulation/DelaunayPointSet.scala]: `IndexedPointSet`

File: spark/src/main/scala/geotrellis/spark/triangulation/BoundaryDelaunay.scala [hunk 6]
```diff
@@ -0,0 +1,426 @@
...
+        case Some(base) => {
+          var e = base
+          do {
+            if (halfEdgeTable.getDest(e) == getDest(tri)) {
+              return Some(e)
+            }
+            e = halfEdgeTable.getNext(e)
+          } while (e != base)
+
+          println("Should never see this")
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/triangulation/BoundaryDelaunay.scala]: Seems like an `IllegalStateException` ?

File: spark/src/main/scala/geotrellis/spark/triangulation/BoundaryDelaunay.scala [hunk 7]
```diff
@@ -0,0 +1,426 @@
+package geotrellis.spark.triangulation
+
+import geotrellis.vector.{Extent, Line, MultiPolygon, Point, Polygon, RobustPredicates}
+import geotrellis.vector.triangulation._
+
+import com.vividsolutions.jts.algorithm.distance.{DistanceToPoint, PointPairDistance}
+import com.vividsolutions.jts.geom.Coordinate
+
+object BoundaryDelaunay {
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/triangulation/BoundaryDelaunay.scala]: It doesn't seem like this class has done anything to deserve to be in `geotrellis.spark`

Comment (jpolchlo) [inline spark/src/main/scala/geotrellis/spark/triangulation/BoundaryDelaunay.scala]: It is of no use outside `geotrellis.spark`.  It exists to support `StitchedDelaunay`, which very much belongs in `spark`

File: spark/src/main/scala/geotrellis/spark/triangulation/StitchedDelaunay.scala [hunk 8]
```diff
@@ -0,0 +1,171 @@
...
+        case 4 => Center
+        case 5 => Right
+        case 6 => BottomLeft
+        case 7 => Bottom
+        case 8 => BottomRight
+      }
+      neighbors(dir)._1.pointSet.getCoordinate(index)
+    }
+
+  def indexToVertex(neighbors: Map[Direction, (DelaunayTriangulation, Extent)])(implicit dummy: DummyImplicit): Int => Coordinate =
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/triangulation/StitchedDelaunay.scala]: The `Direction` class should be duplicated in `geotrellis.raster.tiling.Direction` so we can move this class to `raster` package as well. The version of `Direction` in `spark` should be deprecated.

File: spark/src/main/scala/geotrellis/spark/triangulation/StitchedDelaunay.scala [hunk 9]
```diff
@@ -0,0 +1,171 @@
...
+import geotrellis.raster.triangulation.DelaunayRasterizer
+import geotrellis.spark.buffer.Direction
+import geotrellis.spark.buffer.Direction._
+import geotrellis.vector._
+import geotrellis.vector.io.wkt.WKT
+import geotrellis.vector.triangulation._
+
+import com.vividsolutions.jts.geom.Coordinate
+
+object StitchedDelaunay {
```
Comment (echeipesh) [inline spark/src/main/scala/geotrellis/spark/triangulation/StitchedDelaunay.scala]: This should be moved to `geotrellis.vector.triangulation`

**Pull request merged**